crates/red_knot_python_semantic/src/types.rs

```diff
@@ -959,41 +1024,66 @@ impl<'db> CallOutcome<'db> {
                     };
                     union_builder = union_builder.add(return_ty);
                 }
+                let return_ty = union_builder.build();
                 match not_callable[..] {
-                    [] => {}
-                    [elem] => builder.add_diagnostic(
-                        node,
-                        "call-non-callable",
-                        format_args!(
-                            "Object of type '{}' is not callable (due to union element '{}').",
-                            called_ty.display(db),
-                            elem.display(db),
-                        ),
-                    ),
-                    _ if not_callable.len() == outcomes.len() => builder.add_diagnostic(
-                        node,
-                        "call-non-callable",
-                        format_args!(
-                            "Object of type '{}' is not callable.",
-                            called_ty.display(db)
-                        ),
-                    ),
-                    _ => builder.add_diagnostic(
-                        node,
-                        "call-non-callable",
-                        format_args!(
-                            "Object of type '{}' is not callable (due to union elements {}).",
-                            called_ty.display(db),
-                            not_callable.display(db),
-                        ),
-                    ),
+                    [] => Ok(return_ty),
+                    [elem] => Err(NotCallableError::UnionElement {
+                        not_callable_ty: elem,
+                        called_ty: *called_ty,
+                        return_ty,
+                    }),
+                    _ if not_callable.len() == outcomes.len() => Err(NotCallableError::Union {
+                        called_ty: *called_ty,
+                        return_ty,
+                    }),
+                    _ => Err(NotCallableError::UnionElements {
+                        not_callable_tys: not_callable.into_boxed_slice(),
+                        called_ty: *called_ty,
+                        return_ty,
+                    }),
                 }
-                union_builder.build()
             }
         }
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum NotCallableError<'db> {
+    /// The type is not callable.
+    Type {
+        not_callable_ty: Type<'db>,
+        return_ty: Type<'db>,
+    },
+    /// None of the union elements are callable.
+    Union {
+        called_ty: Type<'db>,
+        return_ty: Type<'db>,
+    },
```

Comment:
I think we should just use the `Type` variant for this case. The default diagnostics are the same, and kind of the point is that a fully-not-callable union shouldn't really be treated differently from any other not-callable type.